src/main/java/graphql/GraphQL.java

```diff
@@ -244,50 +244,108 @@ public ExecutionResult execute(String requestString, String operationName, Objec
+        parseInstrumentation.onEnd(document);
+        return ParseResult.of(document);
+    }
+
+    private List<ValidationError> validate(ExecutionInput executionInput, Document document) {
+        InstrumentationContext<List<ValidationError>> validationCtx = instrumentation.beginValidation(new InstrumentationValidationParameters(executionInput, document));
 
         Validator validator = new Validator();
         List<ValidationError> validationErrors = validator.validateDocument(graphQLSchema, document);
 
         validationCtx.onEnd(validationErrors);
+        return validationErrors;
+    }
 
-        if (validationErrors.size() > 0) {
-            return new ExecutionResultImpl(validationErrors);
-        }
-        ExecutionId executionId = idProvider.provide(requestString, operationName, context);
+    private ExecutionResult execute(ExecutionInput executionInput, Document document) {
+        String requestString = executionInput.getRequestString();
+        String operationName = executionInput.getOperationName();
+        Object context = executionInput.getContext();
+        Object root = executionInput.getRoot();
+        Map<String, Object> arguments = executionInput.getArguments() != null ? executionInput.getArguments() : Collections.emptyMap();
 
         Execution execution = new Execution(queryStrategy, mutationStrategy, subscriptionStrategy, instrumentation);
-        ExecutionResult result = execution.execute(executionId, graphQLSchema, context, root, document, operationName, arguments);
+        ExecutionId executionId = idProvider.provide(requestString, operationName, context);
+        return execution.execute(executionId, graphQLSchema, context, root, document, operationName, arguments);
+    }
+
+    private ExecutionResult toParseFailureExecutionResult(RecognitionException exception) {
+        InvalidSyntaxError invalidSyntaxError = toInvalidSyntaxError(exception);
+        return new ExecutionResultImpl(invalidSyntaxError);
+    }
+
+    private InvalidSyntaxError toInvalidSyntaxError(RecognitionException recognitionException) {
+        SourceLocation sourceLocation = null;
+        if (recognitionException != null) {
+            sourceLocation = new SourceLocation(recognitionException.getOffendingToken().getLine(), recognitionException.getOffendingToken().getCharPositionInLine());
+        }
+        return new InvalidSyntaxError(sourceLocation);
+    }
+
+    private static class ParseResult {
+        private final Document data;
+        private final RecognitionException exception;
 
-        executionCtx.onEnd(result);
+        ParseResult(Document data, RecognitionException exception) {
+            this.data = data;
+            this.exception = exception;
+        }
+
+        boolean isFailure() {
+            return data == null;
+        }
+
+        public Document getData() {
```

Comment:
Can we call this getDocument please

nit: no need for public modifier on private class